kv/util/util.go

```diff
@@ -238,57 +321,65 @@ func watchAndUpdate(
 	getValue getValueFn,
 	update updateFn,
 	validate ValidateFn,
-	defaultValue interface{},
 	logger xlog.Logger,
-) (kv.ValueWatch, error) {
+) error {
 	if store == nil {
-		return nil, errNilStore
+		return errNilStore
 	}
 
-	watch, err := store.Watch(key)
+	var (
+		watch kv.ValueWatch
+		err   error
+	)
+
+	watch, err = store.Watch(key)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("could not establish initial watch: %v", err)
 	}
 
 	go func() {
 		for {
-			_, ok := <-watch.C()
-			if !ok {
-				return
+			if watch == nil {
+				if watch, err = store.Watch(key); err != nil {
+					logger.Errorf("could not recreate etcd watch: %v", err)
+
+					watch = nil
+					time.Sleep(watchRetryInterval)
+					continue
+				}
 			}
-			updateWithKV(getValue, update, validate, key, watch.Get(), defaultValue, logger)
+
+			for range watch.C() {
+				updateWithKV(getValue, update, key, watch.Get(), validate, logger)
+			}
+			watch = nil
 		}
 	}()
 
-	return watch, err
+	return err
 }
 
 func updateWithKV(
 	getValue getValueFn,
 	update updateFn,
-	validate ValidateFn,
 	key string,
 	v kv.Value,
-	defaultValue interface{},
+	validate ValidateFn,
 	logger xlog.Logger,
 ) error {
 	if v == nil {
-		// The key is deleted from kv, use the default value.
-		update(defaultValue)
-		logSetDefault(logger, key, defaultValue)
-		return nil
+		logNilUpdate(logger, key)
```

Comment:
I think we should update default value in this case? this(deletion) should actually be a valid case?

Reply:
Yea, I wasn't sure about this. I feel that we can guard against accidentally deleting a key in kv by not applying an update, but then I can also imagine cases where you will want to watch a key but it need not exist. It seems that currently we only watch keys that we expect to exist, so maybe we should include in the comments that the key should always exist so we don't have to address the case of optional keys here and can instead address when the need arises with a different set of functions for that use case?

Reply:
Not really, there are plenty keys that don't exist, for example when we create a new environment, we usually just copy some keys with custom values to the new environment, otherwise they rely on this case to get the default value.

Reply:
Ahh, ok, if it's customary to use the watches with keys that don't exist, then I'll update. Wasn't aware that was the case.